### PR TITLE
ref: document cloud versioned remotes

### PR DESCRIPTION
**Pull request by pmrowla**
> You may disregard these recommendations if you used the **Edit on GitHub** button from dvc.org to improve a doc in place.

❗ Please read the guidelines in the [Contributing to the Documentation](https://dvc.org/doc/user-guide/contributing/docs) list if you make any substantial changes to the documentation or JS engine.

🐛 Please make sure to mention `Fix #issue` (if applicable) in the description of the PR. This causes GitHub to close it automatically when the PR is merged.

Please choose to [allow us](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to edit your branch when creating the PR.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

**Comment by pmrowla** [comment]
Wasn't really sure where to document the new remote configurations since we don't actually document much about how traditional remotes work either, so I added a new page in `user-guide/data-management/` for now

File: content/docs/command-reference/update.md [hunk 1]
```diff
@@ -38,6 +40,19 @@ to update an imported artifact to a different revision.
...
 
+### Worktree update
+
+When using a
+[worktree](/doc/user-guide/data-management/cloud-versioning#worktree-remotes)
+remote, `dvc update` will update the specified target to match the latest
+version of the corresponding file or directory from the remote storage. If the
+"latest" version of the specified target is a deleted file or an empty
+directory, `dvc update` will fail (in order to avoid potential accidental local
+data loss).
```
Comment (dberenbaum) [inline content/docs/command-reference/update.md:51]: Is this really about accidental data loss? Couldn't the same be said for overwriting local data? Do we need a rationale here?

File: content/docs/user-guide/data-management/cloud-versioning.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+filenames without the use of the DVC CLI (`dvc get --show-url`) or API
+(`dvc.api.get_url()`).
+
+DVC supports the use of cloud object versioning for cases where users prefer to
+retain their original filenames and directory hierarchy in remote storage, in
+exchange for losing the de-duplication and performance benefits of
+content-addressible storage. When cloud versioning is enabled, DVC will store
+files in the remote according to their original directory location and
+filenames. Different versions of a file will then be stored as separate versions
+of the corresponding object in cloud storage.
```
Comment (dberenbaum) [inline content/docs/user-guide/data-management/cloud-versioning.md:19]: > When cloud versioning is enabled, DVC will store
files in the remote according to their original directory location and
filenames. Different versions of a file will then be stored as separate versions
of the corresponding object in cloud storage.

I'd suggest starting Cloud Versioning with this part, and then explain the differences from the default content-addressable storage (possible in its own section or collapsible info blurb).

File: content/docs/command-reference/remote/modify.md [hunk 3]
```diff
@@ -346,6 +346,19 @@ $ dvc push
...
+  cloud versioning features for this S3 remote. This requires that
+  [S3 Versioning](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Versioning.html)
+  be enabled on the specified S3 bucket.
+
+- `worktree` - Use
+  [worktree](/docs/user-guide/data-management/cloud-versioning#worktree-remotes)
+  cloud versioning features for this S3 remote. This requires that
+  [S3 Versioning](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Versioning.html)
+  be enabled on the specified S3 bucket. When both `version_aware` and
+  `worktree` are set, `worktree` takes precedence.
```
Comment (dberenbaum) [inline content/docs/command-reference/remote/modify.md:360]: Does this mean that `worktree` alone is enough to set and that additionally setting `version_aware` is redundant?

File: content/docs/user-guide/data-management/cloud-versioning.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  will retain its original directory structure and filenames, and each version
+  of a file tracked by DVC will be stored as a new version of the corresponding
+  object in cloud storage.
+- `dvc fetch` and `dvc pull` will download the corresponding version of an
+  object from cloud storage.
+
+⚠️ Note that when `version_aware` is in use, DVC does not set DELETE flags on
+objects in cloud storage, and does not make any attempt to ensure that the
+latest version of an object in cloud storage matches the latest version of a
+file in your DVC repository.
```
Comment (dberenbaum) [inline content/docs/user-guide/data-management/cloud-versioning.md:53]:
```suggestion
⚠️ Note that when `version_aware` is in use, DVC does not delete current versions
or restore noncurrent versions of objects in cloud storage, so the current version of an object
in cloud storage may not match the latest version of a file in your DVC repository.
```

Tried to use language that was most common across the cloud versioning docs in:
* https://docs.aws.amazon.com/AmazonS3/latest/userguide/versioning-workflows.html
* https://learn.microsoft.com/en-us/azure/storage/blobs/versioning-overview
* https://cloud.google.com/storage/docs/object-versioning

File: content/docs/user-guide/data-management/cloud-versioning.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  but that do not exist in your current DVC repository workspace.
+- `dvc fetch` and `dvc pull` will download the corresponding version of an
+  object from cloud storage.
+- `dvc update` can be used to update a DVC-tracked file or directory in your
+  current workspace to match the latest version of the corresponding object(s)
+  from cloud storage.
+
+⚠️ Note that setting DELETE flags does not delete any object versions (and does
+not delete any data) from cloud storage, it only means that the "latest" version
+of a given object will show that the object does not exist.
```
Comment (dberenbaum) [inline content/docs/user-guide/data-management/cloud-versioning.md:74]:
```suggestion
⚠️ Note that deleting current versions does not delete any objects (and does
not delete any data) from cloud storage, it only means that the current version
of a given object will show that the object does not exist.
```

File: content/docs/user-guide/data-management/cloud-versioning.md (hunk 4 above)
Comment (dberenbaum) [reply content/docs/user-guide/data-management/cloud-versioning.md:53]: Actually, I would move this down and combine it with the `Worktree remotes` section since it's more useful in comparison.

File: content/docs/user-guide/data-management/cloud-versioning.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  object from cloud storage.
+
+⚠️ Note that when `version_aware` is in use, DVC does not set DELETE flags on
+objects in cloud storage, and does not make any attempt to ensure that the
+latest version of an object in cloud storage matches the latest version of a
+file in your DVC repository.
+
+## Worktree remotes
+
+When the `worktree` option is enabled on a `dvc remote`:
```
Comment (dberenbaum) [inline content/docs/user-guide/data-management/cloud-versioning.md:57]: Can you frame the `Worktree remotes` section in comparison to `Version-aware remotes`, like:

```
When the `worktree` option is enabled on a `dvc remote`, it does the same as `version_aware`, and in addition:
```

Then you could cut the rest of this section down to focus only on the differences instead of a mix of repeated and new info.

File: content/docs/command-reference/update.md (hunk 1 above)
Comment (dberenbaum) [reply content/docs/command-reference/update.md:51]: To clarify: not a blocker.

File: content/docs/user-guide/data-management/cloud-versioning.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  of the remote storage is a mirror of your current local DVC repository
+  workspace. Data in cloud storage will retain its original directory structure
+  and filenames, and each version of a file tracked by DVC will be stored as a
+  new version of the corresponding object in cloud storage. Additionally, DVC
+  will set the DELETE flag on any objects which were present in cloud storage
+  but that do not exist in your current DVC repository workspace.
+- `dvc fetch` and `dvc pull` will download the corresponding version of an
+  object from cloud storage.
+- `dvc update` can be used to update a DVC-tracked file or directory in your
+  current workspace to match the latest version of the corresponding object(s)
```
Comment (dberenbaum) [inline content/docs/user-guide/data-management/cloud-versioning.md:69]:
```suggestion
  current workspace to match the current version of the corresponding object(s)
```

File: content/docs/command-reference/update.md [hunk 8]
```diff
@@ -38,6 +40,19 @@ to update an imported artifact to a different revision.
 $ dvc update --rev master
 ```
 
+### Worktree update
+
+When using a
+[worktree](/doc/user-guide/data-management/cloud-versioning#worktree-remotes)
+remote, `dvc update` will update the specified target to match the latest
+version of the corresponding file or directory from the remote storage. If the
+"latest" version of the specified target is a deleted file or an empty
```
Comment (dberenbaum) [inline content/docs/command-reference/update.md:49]:
```suggestion
remote, `dvc update` will update the specified target to match the current
version of the corresponding file or directory from the remote storage. If the
current version of the specified target is a deleted file or an empty
```

File: content/docs/command-reference/remote/modify.md (hunk 3 above)
Comment (pmrowla) [reply content/docs/command-reference/remote/modify.md:360]: Yes, `worktree = true` now implies `version_aware = true`, so you no longer need to set both. If both `worktree = true` and `version_aware = false` are explicitly set you will get an error

**Comment by dberenbaum** [comment]
@pmrowla Let's get this merged as soon as you're back. A couple more ideas to consider:

* We can add some kind of admon warning that these features are still experimental/in testing and subject to frequent changes.
* Should we update anything from https://github.com/iterative/dvc.org/pull/4142 (link to the docs here? clarification of terms?)

**Comment by pmrowla** [comment]
> @pmrowla Let's get this merged as soon as you're back. A couple more ideas to consider:
> 
> * We can add some kind of admon warning that these features are still experimental/in testing and subject to frequent changes.
> * Should we update anything from [ref: document `import-url` cloud versioning changes #4142](https://github.com/iterative/dvc.org/pull/4142) (link to the docs here? clarification of terms?)

I pushed changes for the current batch of review comments, and took care of these two points as well (for the import-url case I just added the relevant links for now).

It looks like the heroku build is failing but I don't have the permissions to see why, just let me know whatever needs to be fixed (may be something w/my `<admon>` tag usage?).

**Comment by dberenbaum** [comment]
@iterative/websites The build looks okay locally. Could you take a look?

File: content/docs/command-reference/remote/modify.md [hunk 9]
```diff
@@ -346,6 +346,26 @@ $ dvc push
...
 [boto3 docs](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#using-environment-variables)
 
+- `version_aware` - Use
+  [version-aware](/docs/user-guide/data-management/cloud-versioning#version-aware-remotes)
+  cloud versioning features for this S3 remote. Files stored in the remote will
+  retain their original filenames and directory hierarchy, and different
+  versions of files will be stored as separate versions of the corresponding
+  object in the remote. This requires that
+  [S3 Versioning](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Versioning.html)
+  be enabled on the specified S3 bucket.
```
Comment (dberenbaum) [inline content/docs/command-reference/remote/modify.md:356]: Minor: should we call this out with an admon?

File: content/docs/user-guide/data-management/cloud-versioning.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+filenames and directory hierarchy in remote storage (instead of using DVC's
+usual
+[content-addressible storage](/doc/user-guide/project-structure/internal-files#structure-of-the-cache-directory)
+format).
+
+<admon type="warn">
+
+Note that not all DVC functionality is supported when using cloud versioned
+remotes, and using cloud versioning comes with the tradeoff of losing certain
+benefits of content-addressible storage.
```
Comment (dberenbaum) [inline content/docs/user-guide/data-management/cloud-versioning.md:25]:
```suggestion
benefits of content-addressable storage.
```

File: content/docs/command-reference/remote/modify.md (hunk 9 above)
Comment (pmrowla) [reply content/docs/command-reference/remote/modify.md]: moved this into a `tip` admon for each provider

**Review by dberenbaum — APPROVED** [review]

**Review by dberenbaum — APPROVED** [review]

**Comment by jorgeorpinel** [comment]
> we don't actually document much about how traditional remotes work either, so I added a new page in user-guide/data-management/ for now

Note that there are 2 drafts in the works (from me) that attempt to better document [data versioning](https://github.com/iterative/dvc.org/pull/4119) (including cloud versioning) and [remote configuration](https://github.com/iterative/dvc.org/pull/4264), respectively.

**Comment by jorgeorpinel** [comment]
Thanks for https://dvc.org/doc/user-guide/data-management/cloud-versioning BTW, it's a great start.

File: content/docs/user-guide/data-management/cloud-versioning.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+filenames and directory hierarchy in remote storage (instead of using DVC's
+usual
+[content-addressable storage](/doc/user-guide/project-structure/internal-files#structure-of-the-cache-directory)
+format).
+
+<admon type="warn">
+
+Note that not all DVC functionality is supported when using cloud versioned
+remotes, and using cloud versioning comes with the tradeoff of losing certain
+benefits of content-addressable storage.
```
Comment (jorgeorpinel) [inline content/docs/user-guide/data-management/cloud-versioning.md:25]: What specific functionality is not available? The hidden section below only states that "DVC does not provide de-duplication, and certain remote storage performance optimizations will be unavailable".

File: content/docs/user-guide/data-management/cloud-versioning.md [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+  of a file tracked by DVC will be stored as a new version of the corresponding
+  object in cloud storage.
+- `dvc fetch` and `dvc pull` will download the corresponding version of an
+  object from cloud storage.
+
+<admon type="warn">
+
+Note that when `version_aware` is in use, DVC does not delete current versions
+or restore noncurrent versions of objects in cloud storage. So the current
+version of an object in cloud storage may not match the version of a file in your DVC repository.
```
Comment (jorgeorpinel) [inline content/docs/user-guide/data-management/cloud-versioning.md:78]: Can we re-explain this in simpler terms? I'm not sure I get it but it seems key (and the differenve vs. worktree remotes).

> BTW do we have a dummy versioned bucket to play with by any chance?

File: content/docs/user-guide/data-management/cloud-versioning.md (hunk 11 above)
Comment (pmrowla) [reply content/docs/user-guide/data-management/cloud-versioning.md:25]: pretty much anything that isn't explicitly documented as a cloud-versioning-enabled command/feature

But off the top of my head, some of the things that don't work are:

- `dvc import` from a DVC repo that uses cloud versioned remote(s)
- run cache (it works locally but you cannot push/pull it with cloud versioned remotes)
- `gc -c`
- `status -c`
- `exp` sharing (anything that requires `exp push/pull`)
- `dvc push` with revision flags (you cannot `dvc push --all-branches/--all-tags/--all-commits` to cloud versioned remotes, but `dvc pull` will work with the flags)

File: content/docs/user-guide/data-management/cloud-versioning.md (hunk 12 above)
Comment (pmrowla) [reply content/docs/user-guide/data-management/cloud-versioning.md:78]: I'm not really sure there is a simpler way to explain it. The assumption here is that users have at least a basic understanding of how cloud versioning works in S3/azure/GCS.

Deleting a file from a versioned bucket in cloud storage does not delete any data, but it sets a flag that says the file is "deleted" (i.e. it becomes hidden). So if you delete a file on S3, and then ask S3 for the latest version of that file, it will say the file does not exist. But if you want to access the older versions of that file, you can still get them from S3 (as long as you ask for the specific version you want, and not just "latest/current version of file"). Likewise, by default S3 gives you the most recent (current) version of a file unless you request a specific version. So when you look at the bucket in the S3 console, you would only see the latest/current versions of what is in your bucket.

Also, it's important to note that in versioned storage you can only ever add new versions of a file (you cannot revert to an older version so that the old version becomes the current/latest version).  Doing a "revert" operation actually creates a new copy of the older version of the file (and the duplicate costs against your S3 storage quota, even though they are identical).

---

For `worktree`, we tell S3 to "delete" files that are not in the DVC repo (so the latest/current version of a bucket looks like a mirror of the DVC repo). Likewise, for files that do exist in both the DVC repo and the bucket, we make sure the "latest/current" version always matches the DVC repo. If you have someone manually editing files in the bucket, or two people pushing from different branches in the DVC repo, each `dvc push` will overwrite the latest/current version with a new copy of files (that match the repo at the time a user does `dvc push`). (this will result in duplicated data)

For `version_aware`, we don't bother deleting anything, we only care about whether or not the version we care about was pushed at some point in time. Meaning the "latest/current" version of files in your bucket may not actually represent the latest version of your DVC repo. As long as the version of a file from our DVC repo was pushed at some point in time (and that pushed version still exists in the bucket) we won't push duplicate data.

File: content/docs/user-guide/data-management/cloud-versioning.md (hunk 11 above)
Comment (jorgeorpinel) [reply content/docs/user-guide/data-management/cloud-versioning.md:25]: It may be worth it to rephrase this and other texts then, to emphasize that A LOT of features don't work with cloud versioned data remotes.

File: content/docs/user-guide/data-management/cloud-versioning.md (hunk 12 above)
Comment (jorgeorpinel) [reply content/docs/user-guide/data-management/cloud-versioning.md:78]: Yes, there should be a simpler way to explain it, or maybe it's not needed. TBH I don't even know what the point of the admonition is. What's the warning? Esp. if we're assuming the user knows how versioned storage works -- shouldn't it be obvious that DVC can't delete anything (since it can't be done) ?

> whether or not the version we care about was pushed at some point in time. Meaning the "latest/current" version of files in your bucket may not actually represent the latest version of your DVC repo

I think this is closer to what we're trying to emphasize here. Maybe something like:

"Because versioned storage does not allow true deletions or directly restoring old versions [link somewhere], there may be situations where the latest data in your DVC repo does not match what you see as latest versions in the bucket."

It doesn't fully explain how those situations happen but I think that's what we're trying to make a note about.

> p.s. sorry very late reply cc @dberenbaum
